File: backends/java/zappy-application/src/main/java/com/zappymart/application/accounts/FindSignedInCustomer.java
package com.zappymart.application.accounts;

import com.zappymart.application.Visitor;
import com.zappymart.application.ports.CustomerRepository;
import com.zappymart.domain.accounts.Customer;

import java.util.Optional;

public final class FindSignedInCustomer {

    private final CustomerRepository customerRepository;

    public FindSignedInCustomer(CustomerRepository customerRepository) {
        this.customerRepository = customerRepository;
    }

    public Optional<Customer> execute(Visitor visitor) {
        if (!visitor.isSignedIn()) {
            return Optional.empty();
        }
        return customerRepository.byId(visitor.customerId());
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/accounts/FindSignedInCustomer.kt
package nl.zappymart.application.accounts

import nl.zappymart.application.Visitor
import nl.zappymart.application.ports.CustomerRepository
import nl.zappymart.domain.accounts.Customer

class FindSignedInCustomer(private val customers: CustomerRepository) {

    fun execute(visitor: Visitor): Customer? = visitor.customerId?.let { id -> customers.findById(id) }
}
